### PR TITLE
MINOR: Log4j Improvements on Fetcher

### PR DESCRIPTION
**Pull request by guozhangwang**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @hachikuji 

With #8600, I still get various errors (hanging on global restoration, consumer not getting all records in time, etc) every about 170 runs; with this PR on top of #8600 I'm able to run the integration tests 500+ runs without failure.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -258,12 +258,12 @@ public synchronized int sendFetches() {
...
-            RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
-            // We add the node to the set of nodes with pending fetch requests before adding the
-            // listener because the future may have been fulfilled on another thread (e.g. during a
+            // We add the node to the set of nodes with pending fetch requests before sending the
+            // request to the client because the future may have been fulfilled on another thread (e.g. during a
             // disconnection being handled by the heartbeat thread) which will mean the listener
-            // will be invoked synchronously.
+            // will be invoked synchronously, and hence the added id would not be removed anymore.
             this.nodesWithPendingFetchRequests.add(entry.getKey().id());
+            RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:266]: This is the actual fix, others are just log4j improvements I got during the debugging.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -260,7 +260,7 @@ public synchronized int sendFetches() {
             }
             RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
             // We add the node to the set of nodes with pending fetch requests before adding the
-            // listener because the future may have been fulfilled on another thread (e.g. during a
+            // listenerbecause the future may have been fulfilled on another thread (e.g. during a
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I assume this was unintentional.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 3]
```diff
@@ -676,13 +676,15 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
             if (completedFetch.nextFetchOffset == position.offset) {
                 List<ConsumerRecord<K, V>> partRecords = completedFetch.fetchRecords(maxRecords);
 
+                log.trace("Returning fetched records {} at offset {} for assigned partition {}",
+                        partRecords, position, completedFetch.partition);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I feel logging all of the records even at TRACE level will be too much. For example, our system tests often have TRACE enabled. Huge single-line log messages are difficult to consume both visually and in systems like elastic.

**Comment by guozhangwang** [comment]
@mjsax @hachikuji My previous investigation on this issue ended up in the broker-side `DefaultRecordBatch#setLastOffset` which could some time set garbage into the byte buffer (sometimes deterministic but incorrect values, sometimes total random values), and after 36+ hours of chasing this lead I came to the conclusion that it is not really a bug in our code but more likely my IDEA's rabbit holes.. for example yesterday night with rebasing I managed to be able to run 500+ without failure and I peeled off each of the changes I made and it still succeeds, while this morning being on top of trunk #8400 it again fails on 170+ runs, and with all of this observations I'd have to say so long with my past two days and that this is likely not related to any issues on broker or consumer side..

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 3 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Makes sense, it is primarily for my local debugging of an integration test which only sends a couple records. I will only print the num.records instead.

**Review by hachikuji — APPROVED** [review]
LGTM

**Pull request merged**